This protects against the risk of people leaving their VM running and racking up a large bill (for reference, the default VM size here costs $0.79/hour), but unlike the the GCE version of this script, it puts you at risk of losing your work. Deallocating the VM also destroys its disk (while on GCE, shutting down the VM leaves the disk intact. This still costs money, but at a much lower rate). I think before we make a change to auto-deallocate azure VMs, we should change to attaching a separate disk. 

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[scripts/azworker.sh, line 61 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17030#-Kp62M8Iy5HzO1MPvHb5:-Kp62M8Iy5HzO1MPvHb6:b-bk1o6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a1767517f252f3428a11795d9ea734e917ab22/scripts/azworker.sh#L61)):*
> ```Shell
>     rsync -az "$(dirname "${0}")/../build/disable-hyperv-timesync.sh" "${USER}@${FQDN}:bootstrap/"
>     # Copy the azure config (including access tokens).
>     rsync -az ~/.azure "${USER}@${FQDN}:./"
> ```

`~/.azure` also includes a log file, which seems silly to copy (although it looks like it would take a very long time for the log to grow to a problematic size).

Copying full-powered credentials doesn't seem like a great idea, but since we're already using `ssh -A` i guess that ship has sailed, and I don't have good alternatives. 

---

*[scripts/azworker.sh, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17030#-Kp659xb-W1LmzhD8-u-:-Kp659xb-W1LmzhD8-u0:bqsutr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e4a1767517f252f3428a11795d9ea734e917ab22/scripts/azworker.sh#L76)):*
> ```Shell
> 
>     # shellcheck disable=SC2029
>     ssh "${USER}@${FQDN}" "echo '* * * * * /home/radu/bootstrap/autoshutdown.cron.sh 10 az vm deallocate --resource-group \"${RG}\" --name \"${NAME}\"' | crontab -"
> ```

We use the `azure` command elsewhere, not `az` (There are two different azure CLI toolchains. I thought `azure` was the newer one when i wrote this script, but maybe I had it backwards). Add a TODO to switch the rest of this script to the `az` interface. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17030#-:-Kp63SB-2LPIVZ-Ui9XE:b-bmhuku)*
<!-- Sent from Reviewable.io -->

--- a/build/bootstrap/autoshutdown.cron.sh
+++ b/build/bootstrap/autoshutdown.cron.sh
@@ -7,12 +7,13 @@
 
 set -euxo pipefail
 
-if [ "$#" -ne 1 ]; then
-  echo "Usage: $0 <num_periods>"
+if [ -z "$1" ]; then
+  echo "Usage: $0 <num_periods> [shutdown command...]"
   exit 1
 fi
 
 MAX_COUNT="$1"
+shift
 
 if ! [ "$MAX_COUNT" -gt 0 -a "$MAX_COUNT" -lt 10000 ]; then
   echo "Invalid argument '$MAX_COUNT'"
@@ -25,7 +26,8 @@ fi
 FILE=/dev/shm/autoshutdown-count
 COUNT=0
 
-if [ -f /.active ] || w -hs | grep -q pts; then
+
+if [ -f /.active ] || w -hs | grep pts | grep -vq "pts/[0-9]* *tmux"; then
   # Auto-shutdown is disabled (via /.active) or there is a remote session.
   echo 0 > $FILE
   exit 0
@@ -39,6 +41,14 @@ COUNT=$((COUNT+1))
 
 if [ $COUNT -le $MAX_COUNT ]; then
   echo $COUNT > $FILE
-else
+  exit 0
+fi
+
+# Shut down.
+
+if [ -z "$1" ]; then
   /sbin/shutdown -h
+else
+  # Run whatever command was passed on the command line.
+  $@
 fi

--- a/build/bootstrap/install-azure-cli.sh
+++ b/build/bootstrap/install-azure-cli.sh
@@ -1,0 +1,14 @@
+#!/usr/bin/env bash
+#
+# On a Debian/Ubuntu system, installs the Azure CLI.
+
+set -euxo pipefail
+
+# Instructions from https://docs.microsoft.com/en-us/cli/azure/install-azure-cli
+echo "deb [arch=amd64] https://packages.microsoft.com/repos/azure-cli/ wheezy main" | \
+     sudo tee /etc/apt/sources.list.d/azure-cli.list
+
+sudo apt-key adv --keyserver packages.microsoft.com --recv-keys 417A0893
+sudo apt-get install -y apt-transport-https
+sudo apt-get update
+sudo apt-get install -y azure-cli

--- a/scripts/azworker.sh
+++ b/scripts/azworker.sh
@@ -22,6 +22,8 @@ IP=${NAME}
 DOMAIN=cockroach-${NAME}
 FQDN=${DOMAIN}.${LOCATION}.cloudapp.azure.com
 
+# TODO(radu): switch to the newer "az" CLI.
+
 case ${1-} in
   create)
     if azure group show "${RG}" >/dev/null 2>/dev/null; then
@@ -59,6 +61,11 @@ case ${1-} in
     rsync -az "$(dirname "${0}")/../build/disable-hyperv-timesync.sh" "${USER}@${FQDN}:bootstrap/"
     ssh -A "${USER}@${FQDN}" ./bootstrap/bootstrap-debian.sh
     ssh -A "${USER}@${FQDN}" ./bootstrap/disable-hyperv-timesync.sh
+    ssh -A "${USER}@${FQDN}" ./bootstrap/install-azure-cli.sh
+
+    # Copy the azure config (including credentials!).
+    ssh -A "${USER}@${FQDN}" "mkdir .azure"
+    rsync -az ~/.azure/*.json "${USER}@${FQDN}:.azure/"
 
     # Set up tmux configuration (for persistent SSH).
     if [ -e ~/.tmux.conf ]; then
@@ -69,12 +76,8 @@ case ${1-} in
       ssh -A "${USER}@${FQDN}" "echo 'set -g terminal-overrides \"xterm*:XT:smcup@:rmcup@\"' >> .tmux.conf"
     fi
 
-    # TODO(bdarnell): autoshutdown.cron.sh does not work on azure. It
-    # halts the VM, but halting the VM doesn't stop billing. The VM
-    # must instead be "deallocated" with the azure API.
-    # Install automatic shutdown after ten minutes of operation without a
-    # logged in user. To disable this, `sudo touch /.active`.
-    #ssh "${USER}@${FQDN}" "sudo cp bootstrap/autoshutdown.cron.sh /root/; echo '* * * * * /root/autoshutdown.cron.sh 10' | sudo crontab -i -"
+    # shellcheck disable=SC2029
+    ssh "${USER}@${FQDN}" "echo '* * * * * /home/radu/bootstrap/autoshutdown.cron.sh 10 az vm deallocate --resource-group \"${RG}\" --name \"${NAME}\"' | crontab -"
 
     echo "VM now running at ${FQDN}"
     ;;

